File: artlinknote/AnalyticsService.swift
// GOAL: Artist profile analytics, related notes, and series detection.
// Pure Swift, no external dependencies. Operates on [Note] and UserProfile.

import Foundation

// MARK: - ArtistProfile Result
struct ArtistProfile {
    let streak: Int
    let fieldCounts: [Field: Int]
    let scores: SkillScores
    let weeklyStats: [WeeklyStat]
    let totalNotes: Int
}

struct SkillScores {
    let volume: Double      // 기록량 (0-100)
    let aiUsage: Double     // AI활용 (0-100)
    let diversity: Double   // 다양성 (0-100)
    let depth: Double       // 깊이 (0-100)
    let consistency: Double // 꾸준함 (0-100)
}

struct WeeklyStat: Identifiable {
    let id: String // "yyyy-Www"
    let weekStart: Date
    let count: Int
    let fields: Set<Field>
}

// MARK: - Analytics Functions

func computeArtistProfile(notes: [Note], profile: UserProfile) -> ArtistProfile {
    let streak = computeStreak(notes: notes)
    let fieldCounts = computeFieldCounts(notes: notes)
    let scores = computeSkillScores(notes: notes, profile: profile)
    let weeklyStats = computeWeeklyStats(notes: notes)

    return ArtistProfile(
        streak: streak,
        fieldCounts: fieldCounts,
        scores: scores,
        weeklyStats: weeklyStats,
        totalNotes: notes.count
    )
}

/// 오늘부터 거슬러 올라가며 연속 기록일 계산
private func computeStreak(notes: [Note]) -> Int {
    guard !notes.isEmpty else { return 0 }

    let calendar = Calendar.current
    let today = calendar.startOfDay(for: Date())

    // Collect unique days that have notes
    var daysWithNotes = Set<Date>()
    for note in notes {
        let day = calendar.startOfDay(for: note.updatedAt)
        daysWithNotes.insert(day)
    }

    // Count consecutive days from today backwards
    var streak = 0
    var checkDate = today

    while daysWithNotes.contains(checkDate) {
        streak += 1
        guard let previousDay = calendar.date(byAdding: .day, value: -1, to: checkDate) else { break }
        checkDate = previousDay
    }

    return streak
}

private func computeFieldCounts(notes: [Note]) -> [Field: Int] {
    var counts: [Field: Int] = [:]
    for note in notes {
        counts[note.field, default: 0] += 1
    }
    return counts
}

/// 5축 스킬 점수 계산 (각 0-100)
private func computeSkillScores(notes: [Note], profile: UserProfile) -> SkillScores {
    guard !notes.isEmpty else {
        return SkillScores(volume: 0, aiUsage: 0, diversity: 0, depth: 0, consistency: 0)
    }

    // 1. 기록량 (Volume): 총 노트 수 기반, 100개에서 만점
    let volume = min(Double(notes.count) / 100.0 * 100.0, 100.0)

    // 2. AI활용 (AI Usage): aiComment가 있는 노트 비율
    let aiCount = notes.filter { $0.aiComment != nil && !($0.aiComment?.isEmpty ?? true) }.count
    let aiUsage = min(Double(aiCount) / Double(notes.count) * 100.0, 100.0)

    // 3. 다양성 (Diversity): 사용한 field 수 / 전체 field 수
    let uniqueFields = Set(notes.map(\.field))
    let diversity = Double(uniqueFields.count) / Double(Field.allCases.count) * 100.0

    // 4. 깊이 (Depth): 평균 본문 길이 기반, 500자 이상이면 만점
    let avgBodyLength = notes.map(\.body.count).reduce(0, +) / notes.count
    let depth = min(Double(avgBodyLength) / 500.0 * 100.0, 100.0)

    // 5. 꾸준함 (Consistency): 최근 30일 중 노트를 쓴 날의 비율
    let calendar = Calendar.current
    let thirtyDaysAgo = calendar.date(byAdding: .day, value: -30, to: Date()) ?? Date()
    let recentNotes = notes.filter { $0.updatedAt >= thirtyDaysAgo }
    let recentDays = Set(recentNotes.map { calendar.startOfDay(for: $0.updatedAt) })
    let consistency = min(Double(recentDays.count) / 30.0 * 100.0, 100.0)

    return SkillScores(
        volume: volume,
        aiUsage: aiUsage,
        diversity: diversity,
        depth: depth,
        consistency: consistency
    )
}

private func computeWeeklyStats(notes: [Note]) -> [WeeklyStat] {
    let calendar = Calendar.current
    var weekMap: [String: (date: Date, count: Int, fields: Set<Field>)] = [:]

    for note in notes {
        let components = calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: note.updatedAt)
        guard let year = components.yearForWeekOfYear, let week = components.weekOfYear else { continue }
        let key = "\(year)-W\(String(format: "%02d", week))"
        let weekStart = calendar.date(from: components) ?? note.updatedAt

        if var existing = weekMap[key] {
            existing.count += 1
            existing.fields.insert(note.field)
            weekMap[key] = existing
        } else {
            weekMap[key] = (date: weekStart, count: 1, fields: [note.field])
        }
    }

    return weekMap.map { key, value in
        WeeklyStat(id: key, weekStart: value.date, count: value.count, fields: value.fields)
    }.sorted { $0.weekStart < $1.weekStart }
}

// MARK: - Related Notes

/// 관련 노트 추천 (field, tags, series, title 유사도 기반)
func getRelatedNotes(note: Note, allNotes: [Note], maxResults: Int = 5) -> [(note: Note, score: Double)] {
    var scored: [(note: Note, score: Double)] = []

    for candidate in allNotes {
        guard candidate.id != note.id else { continue }
        var score = 0.0

        // Same field: +3
        if candidate.field == note.field {
            score += 3.0
        }

        // Shared tags
        let sharedTags = Set(note.tags).intersection(Set(candidate.tags))
        score += Double(sharedTags.count) * 2.0

        // Same series: +5
        if let series = note.seriesName, !series.isEmpty, candidate.seriesName == series {
            score += 5.0
        }

        // Title similarity (Jaccard on words)
        let noteWords = Set(note.title.lowercased().components(separatedBy: .whitespacesAndNewlines).filter { $0.count >= 2 })
        let candidateWords = Set(candidate.title.lowercased().components(separatedBy: .whitespacesAndNewlines).filter { $0.count >= 2 })
        if !noteWords.isEmpty && !candidateWords.isEmpty {
            let intersection = noteWords.intersection(candidateWords).count
            let union = noteWords.union(candidateWords).count
            let jaccard = Double(intersection) / Double(union)
            score += jaccard * 4.0
        }

        // Recency bonus (more recent = slight boost)
        let daysSinceUpdate = Date().timeIntervalSince(candidate.updatedAt) / 86400
        if daysSinceUpdate < 7 { score += 1.0 }
        else if daysSinceUpdate < 30 { score += 0.5 }

        if score > 0 {
            scored.append((note: candidate, score: score))
        }
    }

    return scored
        .sorted { $0.score > $1.score }
        .prefix(maxResults)
        .map { $0 }
}

// MARK: - Series Detection

/// 시리즈 자동 감지: seriesName이 같은 노트를 그룹핑
func getNoteSeries(allNotes: [Note]) -> [String: [Note]] {
    var seriesMap: [String: [Note]] = [:]

    for note in allNotes {
        if let series = note.seriesName, !series.isEmpty {
            seriesMap[series, default: []].append(note)
        }
    }

    // Also detect implicit series by common title prefixes (e.g. "Act 1:", "Act 2:")
    let unseriedNotes = allNotes.filter { $0.seriesName == nil || $0.seriesName?.isEmpty == true }
    var prefixGroups: [String: [Note]] = [:]

    for note in unseriedNotes {
        // Try to extract a prefix before common separators
        let separators = [":", " -", " –", " —", "#", " |"]
        for sep in separators {
            if let range = note.title.range(of: sep) {
                let prefix = String(note.title[..<range.lowerBound]).trimmed()
                if prefix.count >= 2 && prefix.count <= 30 {
                    prefixGroups[prefix, default: []].append(note)
                }
                break
            }
        }
    }

    // Only keep prefix groups with 2+ notes
    for (prefix, notes) in prefixGroups where notes.count >= 2 {
        let key = "\(prefix) (auto)"
        seriesMap[key] = notes
    }

    // Sort notes within each series by updatedAt
    for key in seriesMap.keys {
        seriesMap[key]?.sort { $0.updatedAt < $1.updatedAt }
    }

    return seriesMap
}

File: src/services/analyticsService.js
/**
 * analyticsService.js
 *
 * Artist profile analytics, related notes recommendation, and series detection
 * for Artlink React Native.
 *
 * Pure JavaScript, no external dependencies. Operates on note arrays
 * and user profile objects.
 */

import { FIELD_KEYS, startOfDay, getISOWeek } from '../utils/helpers';

// ---------------------------------------------------------------------------
// ArtistProfile shape (returned by computeArtistProfile)
// ---------------------------------------------------------------------------
//
// {
//   streak:      number,       // consecutive days with notes from today
//   fieldCounts: { acting: n, music: n, ... },
//   scores: {
//     volume:      number,  // 0-100  recording volume
//     aiUsage:     number,  // 0-100  AI utilization rate
//     diversity:   number,  // 0-100  field diversity
//     depth:       number,  // 0-100  average body length
//     consistency: number,  // 0-100  recent 30-day activity
//   },
//   weeklyStats: [ { id, weekStart, count, fields } ],
//   totalNotes:  number,
// }

// ---------------------------------------------------------------------------
// computeStreak
// ---------------------------------------------------------------------------

/**
 * Count consecutive days with at least one note, going backwards from today.
 *
 * @param {Array} notes - Array of note objects with `updatedAt` dates
 * @returns {number}
 */
function computeStreak(notes) {
  if (notes.length === 0) return 0;

  const today = startOfDay(new Date());

  // Collect unique day timestamps
  const daysWithNotes = new Set();
  for (const note of notes) {
    const day = startOfDay(new Date(note.updatedAt)).getTime();
    daysWithNotes.add(day);
  }

  let streak = 0;
  let checkDate = today.getTime();
  const ONE_DAY = 86400000;

  while (daysWithNotes.has(checkDate)) {
    streak += 1;
    checkDate -= ONE_DAY;
  }

  return streak;
}

// ---------------------------------------------------------------------------
// computeFieldCounts
// ---------------------------------------------------------------------------

/**
 * Count notes per field.
 *
 * @param {Array} notes
 * @returns {Object} e.g. { acting: 5, music: 3 }
 */
function computeFieldCounts(notes) {
  const counts = {};
  for (const note of notes) {
    counts[note.field] = (counts[note.field] || 0) + 1;
  }
  return counts;
}

// ---------------------------------------------------------------------------
// computeSkillScores  (5-axis, each 0-100)
// ---------------------------------------------------------------------------

/**
 * Compute 5-axis skill scores.
 *
 * 1. Volume (기록량)     -- total notes / 100 * 100, capped at 100
 * 2. AI Usage (AI활용)   -- notes with aiComment / total * 100
 * 3. Diversity (다양성)  -- unique fields used / total field count * 100
 * 4. Depth (깊이)        -- avg body length / 500 * 100, capped at 100
 * 5. Consistency (꾸준함) -- days with notes in last 30 days / 30 * 100
 *
 * @param {Array}  notes
 * @param {Object} _profile  (reserved for future extensions)
 * @returns {{ volume, aiUsage, diversity, depth, consistency }}
 */
function computeSkillScores(notes, _profile) {
  if (notes.length === 0) {
    return { volume: 0, aiUsage: 0, diversity: 0, depth: 0, consistency: 0 };
  }

  // 1. Volume
  const volume = Math.min((notes.length / 100) * 100, 100);

  // 2. AI Usage
  const aiCount = notes.filter(
    (n) => n.aiComment && n.aiComment.trim().length > 0
  ).length;
  const aiUsage = Math.min((aiCount / notes.length) * 100, 100);

  // 3. Diversity
  const uniqueFields = new Set(notes.map((n) => n.field));
  const totalFieldTypes = FIELD_KEYS.length; // 6
  const diversity = (uniqueFields.size / totalFieldTypes) * 100;

  // 4. Depth
  const totalBodyLength = notes.reduce((sum, n) => sum + (n.body || '').length, 0);
  const avgBodyLength = totalBodyLength / notes.length;
  const depth = Math.min((avgBodyLength / 500) * 100, 100);

  // 5. Consistency (last 30 days)
  const now = new Date();
  const thirtyDaysAgo = new Date(now.getTime() - 30 * 86400000);
  const recentNotes = notes.filter(
    (n) => new Date(n.updatedAt) >= thirtyDaysAgo
  );
  const recentDays = new Set(
    recentNotes.map((n) => startOfDay(new Date(n.updatedAt)).getTime())
  );
  const consistency = Math.min((recentDays.size / 30) * 100, 100);

  return { volume, aiUsage, diversity, depth, consistency };
}

// ---------------------------------------------------------------------------
// computeWeeklyStats
// ---------------------------------------------------------------------------

/**
 * Aggregate notes into weekly buckets with counts and field sets.
 *
 * @param {Array} notes
 * @returns {Array<{ id: string, weekStart: Date, count: number, fields: Set<string> }>}
 *   Sorted by weekStart ascending.
 */
function computeWeeklyStats(notes) {
  const weekMap = {};

  for (const note of notes) {
    const { key, weekStart } = getISOWeek(new Date(note.updatedAt));

    if (weekMap[key]) {
      weekMap[key].count += 1;
      weekMap[key].fields.add(note.field);
    } else {
      weekMap[key] = {
        id: key,
        weekStart,
        count: 1,
        fields: new Set([note.field]),
      };
    }
  }

  return Object.values(weekMap).sort(
    (a, b) => a.weekStart - b.weekStart
  );
}

// ---------------------------------------------------------------------------
// computeArtistProfile  (main public function)
// ---------------------------------------------------------------------------

/**
 * Compute the full artist profile analytics object.
 *
 * @param {Array}  notes       - All notes
 * @param {Object} userProfile - User profile ({ name, roleModels, ... })
 * @returns {Object} ArtistProfile
 */
export function computeArtistProfile(notes = [], userProfile = {}) {
  const streak = computeStreak(notes);
  const fieldCounts = computeFieldCounts(notes);
  const scores = computeSkillScores(notes, userProfile);
  const weeklyStats = computeWeeklyStats(notes);

  return {
    streak,
    fieldCounts,
    scores,
    weeklyStats,
    totalNotes: notes.length,
  };
}

// ---------------------------------------------------------------------------
// getRelatedNotes
// ---------------------------------------------------------------------------

/**
 * Recommend related notes based on field, tags, series, and title similarity.
 *
 * Scoring:
 *   - Same field:          +3
 *   - Each shared tag:     +2
 *   - Same series:         +5
 *   - Title similarity:    Jaccard on words * 4
 *   - Recency (< 7 days):  +1 ;  (< 30 days): +0.5
 *
 * @param {Object} note       - The reference note
 * @param {Array}  allNotes   - All notes to search through
 * @param {number} maxResults - Maximum related notes to return (default 5)
 * @returns {Array<{ note: Object, score: number }>}  Sorted by score descending
 */
export function getRelatedNotes(note, allNotes, maxResults = 5) {
  const scored = [];

  for (const candidate of allNotes) {
    if (candidate.id === note.id) continue;

    let score = 0;

    // Same field: +3
    if (candidate.field === note.field) {
      score += 3.0;
    }

    // Shared tags: +2 each
    const noteTags = new Set(note.tags || []);
    const candidateTags = new Set(candidate.tags || []);
    for (const tag of noteTags) {
      if (candidateTags.has(tag)) {
        score += 2.0;
      }
    }

    // Same series: +5
    if (
      note.seriesName &&
      note.seriesName.trim().length > 0 &&
      candidate.seriesName === note.seriesName
    ) {
      score += 5.0;
    }

    // Title similarity: Jaccard on words * 4
    const noteWords = new Set(
      (note.title || '')
        .toLowerCase()
        .split(/\s+/)
        .filter((w) => w.length >= 2)
    );
    const candidateWords = new Set(
      (candidate.title || '')
        .toLowerCase()
        .split(/\s+/)
        .filter((w) => w.length >= 2)
    );

    if (noteWords.size > 0 && candidateWords.size > 0) {
      let intersectionSize = 0;
      for (const w of noteWords) {
        if (candidateWords.has(w)) intersectionSize++;
      }
      const unionSize = new Set([...noteWords, ...candidateWords]).size;
      const jaccard = intersectionSize / unionSize;
      score += jaccard * 4.0;
    }

    // Recency bonus
    const daysSinceUpdate =
      (Date.now() - new Date(candidate.updatedAt).getTime()) / 86400000;
    if (daysSinceUpdate < 7) {
      score += 1.0;
    } else if (daysSinceUpdate < 30) {
      score += 0.5;
    }

    if (score > 0) {
      scored.push({ note: candidate, score });
    }
  }

  return scored
    .sort((a, b) => b.score - a.score)
    .slice(0, maxResults);
}

// ---------------------------------------------------------------------------
// getNoteSeries
// ---------------------------------------------------------------------------

/**
 * Group notes into series.
 *
 * Two detection modes:
 *   1. Explicit: notes with matching `seriesName`
 *   2. Implicit: notes without seriesName whose titles share a common prefix
 *      before separators (: - -- | #). Implicit groups require 2+ notes
 *      and are labeled with "(auto)" suffix.
 *
 * @param {Array} allNotes
 * @returns {Object<string, Array>}  Map of seriesName -> sorted notes array
 */
export function getNoteSeries(allNotes) {
  const seriesMap = {};

  // 1. Explicit series
  for (const note of allNotes) {
    if (note.seriesName && note.seriesName.trim().length > 0) {
      if (!seriesMap[note.seriesName]) {
        seriesMap[note.seriesName] = [];
      }
      seriesMap[note.seriesName].push(note);
    }
  }

  // 2. Implicit series by common title prefixes
  const unseriedNotes = allNotes.filter(
    (n) => !n.seriesName || n.seriesName.trim().length === 0
  );

  const separators = [':', ' -', ' \u2013', ' \u2014', '#', ' |'];
  const prefixGroups = {};

  for (const note of unseriedNotes) {
    for (const sep of separators) {
      const idx = note.title.indexOf(sep);
      if (idx !== -1) {
        const prefix = note.title.substring(0, idx).trim();
        if (prefix.length >= 2 && prefix.length <= 30) {
          if (!prefixGroups[prefix]) {
            prefixGroups[prefix] = [];
          }
          prefixGroups[prefix].push(note);
        }
        break; // only match first separator
      }
    }
  }

  // Only keep prefix groups with 2+ notes
  for (const [prefix, notes] of Object.entries(prefixGroups)) {
    if (notes.length >= 2) {
      const key = `${prefix} (auto)`;
      seriesMap[key] = notes;
    }
  }

  // Sort notes within each series by updatedAt ascending
  for (const key of Object.keys(seriesMap)) {
    seriesMap[key].sort(
      (a, b) => new Date(a.updatedAt) - new Date(b.updatedAt)
    );
  }

  return seriesMap;
}

// ---------------------------------------------------------------------------
// Monthly aggregation helper
// ---------------------------------------------------------------------------

/**
 * Aggregate notes into monthly buckets.
 *
 * @param {Array} notes
 * @returns {Array<{ month: string, count: number, fields: Set<string> }>}
 *   month format: "yyyy-MM", sorted ascending.
 */
export function computeMonthlyStats(notes) {
  const monthMap = {};

  for (const note of notes) {
    const d = new Date(note.updatedAt);
    const key = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`;

    if (monthMap[key]) {
      monthMap[key].count += 1;
      monthMap[key].fields.add(note.field);
    } else {
      monthMap[key] = {
        month: key,
        count: 1,
        fields: new Set([note.field]),
      };
    }
  }

  return Object.values(monthMap).sort((a, b) =>
    a.month.localeCompare(b.month)
  );
}
